transports/tcp/src/lib.rs

```diff
@@ -353,10 +352,63 @@ enum Addresses {
 
 type Buffer = VecDeque<ListenerEvent<FutureResult<TcpTransStream, io::Error>>>;
 
+/// Incoming connection stream which pauses after errors.
+#[derive(Debug)]
+struct Listener<S> {
+    /// The incoming connections.
+    stream: S,
+    /// The current pause if any.
+    pause: Option<Delay>,
+    /// How long to pause after an error.
+    duration: Duration
+}
+
+impl<S> Listener<S>
+where
+    S: Stream,
+    S::Error: std::fmt::Display
+{
+    fn new(stream: S, duration: Duration) -> Self {
+        Listener { stream, pause: None, duration }
+    }
+}
+
+impl<S> Stream for Listener<S>
+where
+    S: Stream,
+    S::Error: std::fmt::Display
+{
+    type Item = S::Item;
+    type Error = S::Error;
+
+    /// Polls for incoming connections, pausing if an error is encountered.
+    fn poll(&mut self) -> Poll<Option<S::Item>, S::Error> {
+        if let Some(mut delay) = self.pause.take() {
```

Comment:
We could use `as_mut()` here in order to not extract and push again the timer all the time.

Reply:
I think `as_mut` saves at most one move per pause. Normally, `pause` is `None` and `Delay` would only notify the task when the time is up.